lib/Sema/CSSimplify.cpp

```diff
@@ -2945,14 +2945,99 @@ bool ConstraintSystem::hasPreconcurrencyCallee(
   return calleeOverload->choice.getDecl()->preconcurrency();
 }
 
+namespace {
+  /// Classifies a thrown error kind as Never, a specific type, or 'any Error'.
+  enum class ThrownErrorKind {
+    Never,
+    Specific,
+    AnyError,
+  };
+
+  ThrownErrorKind getThrownErrorKind(Type type) {
+    if (type->isNever())
+      return ThrownErrorKind::Never;
+
+    if (type->isExistentialType()) {
+      Type anyError = type->getASTContext().getErrorExistentialType();
+      if (anyError->isEqual(type))
+        return ThrownErrorKind::AnyError;
+    }
+
+    return ThrownErrorKind::Specific;
+  }
+}
+
 ConstraintSystem::TypeMatchResult
 ConstraintSystem::matchFunctionTypes(FunctionType *func1, FunctionType *func2,
                                      ConstraintKind kind, TypeMatchOptions flags,
                                      ConstraintLocatorBuilder locator) {
-  // A non-throwing function can be a subtype of a throwing function.
-  if (func1->isThrowing() != func2->isThrowing()) {
-    // Cannot drop 'throws'.
-    if (func1->isThrowing() || kind < ConstraintKind::Subtype) {
+  // A function type that throws the error type E1 is a subtype of a function
+  // that throws error type E2 when E1 is a subtype of E2. For the purpose
+  // of this comparison, a non-throwing function has thrown error type 'Never',
+  // and an untyped throwing function has thrown error type 'any Error'.
+  Type neverType = getASTContext().getNeverType();
+  Type thrownError1 = func1->getEffectiveThrownInterfaceType().value_or(neverType);
+  Type thrownError2 = func2->getEffectiveThrownInterfaceType().value_or(neverType);
+  if (thrownError1 && thrownError2 && !thrownError1->isEqual(thrownError2)) {
```

Comment:
Could you please package this logic into `matchFunctionThrowsAttributes` or something like this and return `TypeMatchResult` from that?